### PR TITLE
[kube-prometheus-stack] Add upgrade path from stable/prometheus-operator

### PR DESCRIPTION
**Pull request by fktkrt**
#### What this PR does / why we need it:

I had to migrate our stable/prometheus-operator to the new chart while keeping the existing PersistentVolume.
To do this I had to perform the steps provided below. After the procedure I was able to access the old time-series data stored in the PV.

#### Which issue this PR fixes
  - fixes https://github.com/prometheus-community/helm-charts/issues/84

#### Special notes for your reviewer:
This guide covers a fresh installation of the new chart, so not persisted configurations (e.g. in values.yaml, custom dashboards) can be lost after the process is finished. 

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/kube-prometheus-stack/README.md
Comment (gkarthiks) [inline charts/kube-prometheus-stack/README.md:226]: Thanks for the content for making an upgrade @fktkrt. How about making few generalized description below? 
/cc @scottrigby 

## Migration steps from stable/prometheus-operator chart

If the **prometheus-operator** values are compatible with the new **kube-prometheus-stack** chart, please follow the below steps for migration

> The guide presumes that chart is deployed in `monitoring` namespace and the deployments are running there. If in other namespace, please replace the `monitoring` to the deployed namespace.

1. Patch the PersistenceVolume created/used by the prometheus-operator chart to `Retain` claim policy:

```bash
kubectl patch pv/<PersistenceVolume name> -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}' -n monitoring
```

**Note:** To execute the above command, the user must have a cluster wide permission. Please refer [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)

2. Uninstall the **prometheus-operator** release, and verify PV become Released.

3. Remove current `spec.claimRef` values to change the PV's status from Released to Available

```bash
kubectl patch pv/<pv_name> --type json -p $'- op: remove\n path: /spec/claimRef'
```

**Note:** To execute the above command, the user must have a cluster wide permission. Please refer [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)

After these steps, proceed to a fresh **kube-prometheus-stack** installation and make sure the current release of `kube-prometheus-stack` matching the `volumeClaimTemplate` values in the `values.yaml`.

The new release should now re-attach your previously released PV with its content.

Comment (fktkrt) [reply charts/kube-prometheus-stack/README.md:230]: Thanks, I was planning to clean it up as well in the morning.

**Review by gkarthiks — COMMENTED** [review]
lgtm, let wait for one more reviewer.

**Comment by scottrigby** [comment]
Thanks 😊 This is great if a user needs to upgrade the release name or namespace.

But given that users can do a full name override to match their existing release, would you be willing to add a similar scenario for retaining PVs for an in-place upgrade?

Also is it clear to everyone what data is being persisted here, and how exactly to do that? Users can persist the grafana subchart user account and dashboard data as well, right?

**Comment by scottrigby** [comment]
Alternatively, we could merge this as-is, and allow other members of the community to document how to do so for an in-place upgrade.￼ WDYT?

**Comment by gkarthiks** [comment]
@scottrigby I like your second option. Merge this as-is and support the @fktkrt for his contribution and others (or may be @fktkrt would be interested as well) might be interested in contributing the other documentation.

**Comment by scottrigby** [comment]
Before this is merged, let's have someone try these steps to ensure they work properly. I'm not sure when I'll get to that but once that's verified we can merge this. 

I created https://github.com/prometheus-community/helm-charts/issues/121 to make sure we don't forget to address scenarios and details I asked about above.

**Comment by fktkrt** [comment]
Just performed it in another cluster, added a missing step and cleaned up the second patch syntax.

**Review by scottrigby — APPROVED** [review]
@fktkrt thanks this is a good start 👍  Looking forward to more community documentation contributions on this in #121

**Comment by scottrigby** [comment]
@gkarthiks I'd merge this but your approval is now stale since more commits were pushed. Feel free to merge after you review the changes

File: charts/kube-prometheus-stack/Chart.yaml
Comment (scottrigby) [inline charts/kube-prometheus-stack/Chart.yaml:20]: You'll need to bump the version again since https://github.com/prometheus-community/helm-charts/pull/109 was just merged

Comment (scottrigby) [inline charts/kube-prometheus-stack/Chart.yaml:20]: You'll need to bump the version again since https://github.com/prometheus-community/helm-charts/pull/109 was just merged

**Review by scottrigby — DISMISSED** [review]

**Comment by Vandersteen** [comment]
Would you be so kind to expand on:

> After these steps, proceed to a fresh kube-prometheus-stack installation and make sure the current release of kube-prometheus-stack matching the volumeClaimTemplate values in the values.yaml.

I got for example these:

```
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM
pvc-7a63cb34-9495-4c14-aba7-4a8758790cd2   128Gi      RWO            Retain           Bound    monitoring/prometheus-prom-prometheus-operator-prometheus-db-prometheus-prom-prometheus-operator-prometheus-1
pvc-9f5412d6-54c5-11ea-b253-3271bc13e8b3   10Gi       RWX            Retain           Bound    monitoring/prom-grafana
pvc-a6803abe-54c5-11ea-b253-3271bc13e8b3   5Gi        RWO            Retain           Bound    monitoring/alertmanager-prom-prometheus-operator-alertmanager-db-alertmanager-prom-prometheus-operator-alertmanager-0
pvc-a682c5a7-54c5-11ea-b253-3271bc13e8b3   5Gi        RWO            Retain           Bound    monitoring/alertmanager-prom-prometheus-operator-alertmanager-db-alertmanager-prom-prometheus-operator-alertmanager-1 
pvc-ac792a14-54c5-11ea-b253-3271bc13e8b3   128Gi      RWO            Retain           Bound    monitoring/prometheus-prom-prometheus-operator-prometheus-db-prometheus-prom-prometheus-operator-prometheus-0 
``` 

And in my (current) values.yaml file:

```
    ## Prometheus StorageSpec for persistent data
    ## ref: https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/storage.md
    ##
    storageSpec:
      volumeClaimTemplate:
        spec:
          accessModes: [ "ReadWriteOnce" ]
          storageClassName: "managed-premium"
          resources:
            requests:
              storage: 128Gi
```

What should I change ?

**Comment by Vandersteen** [comment]
> Would you be so kind to expand on:
> 
> > After these steps, proceed to a fresh kube-prometheus-stack installation and make sure the current release of kube-prometheus-stack matching the volumeClaimTemplate values in the values.yaml.
> 
> I got for example these:
> 
> ```
> NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM
> pvc-7a63cb34-9495-4c14-aba7-4a8758790cd2   128Gi      RWO            Retain           Bound    monitoring/prometheus-prom-prometheus-operator-prometheus-db-prometheus-prom-prometheus-operator-prometheus-1
> pvc-9f5412d6-54c5-11ea-b253-3271bc13e8b3   10Gi       RWX            Retain           Bound    monitoring/prom-grafana
> pvc-a6803abe-54c5-11ea-b253-3271bc13e8b3   5Gi        RWO            Retain           Bound    monitoring/alertmanager-prom-prometheus-operator-alertmanager-db-alertmanager-prom-prometheus-operator-alertmanager-0
> pvc-a682c5a7-54c5-11ea-b253-3271bc13e8b3   5Gi        RWO            Retain           Bound    monitoring/alertmanager-prom-prometheus-operator-alertmanager-db-alertmanager-prom-prometheus-operator-alertmanager-1 
> pvc-ac792a14-54c5-11ea-b253-3271bc13e8b3   128Gi      RWO            Retain           Bound    monitoring/prometheus-prom-prometheus-operator-prometheus-db-prometheus-prom-prometheus-operator-prometheus-0 
> ```
> 
> And in my (current) values.yaml file:
> 
> ```
>     ## Prometheus StorageSpec for persistent data
>     ## ref: https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/storage.md
>     ##
>     storageSpec:
>       volumeClaimTemplate:
>         spec:
>           accessModes: [ "ReadWriteOnce" ]
>           storageClassName: "managed-premium"
>           resources:
>             requests:
>               storage: 128Gi
> ```
> 
> What should I change ?

Oh I see, kubernetes will match the PV's that have the same size / type / ... and are unclaimed

**Comment by fktkrt** [comment]
> Oh I see, kubernetes will match the PV's that have the same size / type / ... and are unclaimed

Yes, that is right. Also this PR only covers the case of 1 Prometheus PV, but the same applies to multiple ones. 
Feel free to reach out if you have further questions.

**Comment by calvinbui** [comment]
A missing step is to delete a left-over service in `kube-system` that monitors kubelets. This is only if you renamed the Helm chart, which is probably true since the chart name has changed from `prometheus-operator` to `kube-prometheus-stack` unless the user is overriding it. [See this comment](https://github.com/kubernetes-monitoring/kubernetes-mixin/issues/392#issuecomment-694610596). I deleted the older one.

image

On GKE I migrated successfully without any issues.

On EKS, the PVC/Pods may be in a different AWS AZ so it creates a new PV. This can be worked around by setting `prometheus.prometheusSpec.nodeSelector` to the same AZ:

```
nodeSelector:
  failure-domain.beta.kubernetes.io/zone: ap-southeast-2a
```

**Comment by fktkrt** [comment]
> A missing step is to delete a left-over service in `kube-system` that monitors kubelets. This is only if you renamed the Helm chart, which is probably true since the chart name has changed from `prometheus-operator` to `kube-prometheus-stack` unless the user is overriding it. [See this comment](https://github.com/kubernetes-monitoring/kubernetes-mixin/issues/392#issuecomment-694610596). I deleted the older one.

Thanks, I've just encountered this as well. Pushed a new commit to include this step.

**Comment by cablespaghetti** [comment]
Tried this on my environment. Unfortunately where our alertmanagers also have PVs it's complete luck whether the right pods gets the right PVs assigned to them. I got around this by deleting the alertmanager PVCs and PVs, then editing the ClaimRef on the prometheus ones to something like:

```
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: prometheus-prometheus-kube-prometheus-prometheus-db-prometheus-prometheus-kube-prometheus-prometheus-1
    namespace: monitoring
```

**Comment by fktkrt** [comment]
> Tried this on my environment. Unfortunately where our alertmanagers also have PVs it's complete luck whether the right pods get the right PVs assigned to them. I got around this by deleting the alertmanager PVCs and PVs, then editing the ClaimRef on the prometheus ones to something like:
> 
> ```
>   claimRef:
>     apiVersion: v1
>     kind: PersistentVolumeClaim
>     name: prometheus-prometheus-kube-prometheus-prometheus-db-prometheus-prometheus-kube-prometheus-prometheus-1
>     namespace: monitoring
> ```

I am wondering if this can be done with component specific labels in the template.

**Comment by gkarthiks** [comment]
@fktkrt regarding this https://github.com/prometheus-community/helm-charts/pull/119#issuecomment-697919685, if we had the selector labels as-is this would work.

**Comment by fktkrt** [comment]
@gkarthiks added these instructions as well.

**Comment by scottrigby** [comment]
I'm not going to be able to test these scenarios personally (not without clear step-by-step testing instructions), as I'm focused on other priorities over the next few weeks.

Once a few of you review, (please test), and approve this PR, I think instructions like this would be very helpful to end users moving from `stable/prometheus-operator` 🙂

File: charts/kube-prometheus-stack/README.md
Comment (desaintmartin) [inline charts/kube-prometheus-stack/README.md:230]: Could some other unrelated PVC, created in between deletion of old PVC and creation of new one, take control of this PV?
Usually, when I do this kind of migration, I delete the old PVC, create the new one with correct PV reference, and edit the PV to point to the new PV.

**Comment by m42u** [comment]
Thanks a lot for this migration plan!
Just tested it and I'd like to share some feedback (on Kubernetes `v1.16.13` using HELM `v2.16.10`):
- You don't need to specify the namespace when running the `kubectl patch` command to patch the PV as PVs are not namespaced
- I had to delete all CRDs related to `*.monitoring.coreos.com` to install `kube-prometheus-stack` otherwise I got errors as CRDs already exists
  - By deleting all CRDs related to `*.monitoring.coreos.com`, it obviously deleted my `ServiceMonitor`s
  - Is there a way to install `kube-prometheus-stack` and skip the CRDs creation?
- I manually deleted the `prometheus-operator-kubelet` `Endpoints` alongside the `prometheus-operator-kubelet` service in the `kube-system` namespace
- PVs were re-attached correctly, but as @desaintmartin mentioned, some other unrelated PVC created in between deletion of old PVC and creation of new one can probably take control of this PV

Thanks!

**Comment by fktkrt** [comment]
> Thanks a lot for this migration plan!
> Just tested it and I'd like to share some feedback (on Kubernetes `v1.16.13` using HELM `v2.16.10`):
> 
>     * You don't need to specify the namespace when running the `kubectl patch` command to patch the PV as PVs are not namespaced
Your're right, only PersistentVolumeClaims are namespaced objects.

>     * I had to delete all CRDs related to `*.monitoring.coreos.com` to install `kube-prometheus-stack` otherwise I got errors as CRDs already exists
>       
>       * By deleting all CRDs related to `*.monitoring.coreos.com`, it obviously deleted my `ServiceMonitor`s
>       * Is there a way to install `kube-prometheus-stack` and skip the CRDs creation?
Yes, you can set `prometheusOperator.createCustomResource` to `false` to disable CRD provisioning.

>     * I manually deleted the `prometheus-operator-kubelet` `Endpoints` alongside the `prometheus-operator-kubelet` service in the `kube-system` namespace
I will include this.

>     * PVs were re-attached correctly, but as @desaintmartin mentioned, some other unrelated PVC created in between deletion of old PVC and creation of new one can probably take control of this PV
True, I AFAIK there's a `volumeName` under `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.spec` that might help, but this needs validation.

>     * There is a missing zone identifier in your example: `failure-domain.beta.kubernetes.io/zone: east-west-1` (`east-west-1a` for example)
Nice catch, thanks!
> 
> Thanks!

@m42u Thanks for the feedback, I will address these and upgrade the PR.

**Comment by Vandersteen** [comment]
> * I had to delete all CRDs related to `*.monitoring.coreos.com` to install `kube-prometheus-stack` otherwise I got errors as CRDs already exists
>   
>   * By deleting all CRDs related to `*.monitoring.coreos.com`, it obviously deleted my `ServiceMonitor`s
>   * Is there a way to install `kube-prometheus-stack` and skip the CRDs creation?

I had this issue too, you can set `prometheusOperator.createCustomResource` to `false`
However in my case it complained that it could not find some CRD's.

I ran the following manually:

```
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_alertmanagers.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_podmonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
```
I then could proceed to install it using createCustomResource: false without losing the servicemonitors

**Comment by m42u** [comment]
> I had this issue too, you can set `prometheusOperator.createCustomResource` to `false`
> However in my case it complained that it could not find some CRD's.
> 
> I ran the following manually:
> 
> ```
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_alertmanagers.yaml
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_podmonitors.yaml
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_prometheusrules.yaml
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml
> kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/release-0.38/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
> ```
> 
> I then could proceed to install it using createCustomResource: false without losing the servicemonitors

FYI, I didn't need to run these commands by upgrading from `prometheus-operator` `v9.3.1` to `kube-prometheus-stack` `v9.4.4`.
I guess some `CRD`s were missing on you cluster.

Everything works as expected on my side using `prometheusOperator.createCustomResource: false`

**Comment by fktkrt** [comment]
@m42u Please see my recent changes to this PR based on your comments.
I've omitted deleting the endpoint, because that should be taken care of when removing the service itself.

> ... there's a `volumeName` under `prometheus.prometheusSpec.storageSpec.volumeClaimTemplate.spec` that might help, but this needs validation.
@desaintmartin, could you maybe look into this option to see whether this solves your comment or not?

**Comment by flouthoc** [comment]
@fktkrt Path requires me to uninstall previous chart release wouldn't this cause any downtimes ?

**Comment by fktkrt** [comment]
> @fktkrt Path requires me to uninstall previous chart release wouldn't this cause any downtimes ?

@flouthoc Yes, between the old is removed and the new is installed (1-2m), you won't have metrics.
If you are seeking a zero-downtime way to upgrade, then helm's`fullnameOverride` option might be a good fit as others have pointed out, although I don't have any personal experience with it (in terms of PV/PVCs at least).

Do you maybe want to give it a try? :)

**Comment by flouthoc** [comment]
@fktkrt Thanks i guess 1-2m wont do any damage. I am performing a drill on dummy cluster to record delta for downtime, I guess after that we are good to go. Any reason why is this guide not merged to master yet ?

**Comment by fktkrt** [comment]
> @fktkrt Thanks i guess 1-2m wont do any damage. I am performing a drill on dummy cluster to record delta for downtime, I guess after that we are good to go. Any reason why is this guide not merged to master yet ?

@scottrigby said that first let's wait for a few more successful migration report then we can merge this.
So @flouthoc, can you report one more for us? :)

**Comment by shaikatz** [comment]
I just did a migration from JSONNET version of kube-prometheus to the helm chart, using those instructions more or less, and it worked perfectly, I think that should get merged :-)

File: charts/kube-prometheus-stack/README.md
Comment (fktkrt) [reply charts/kube-prometheus-stack/README.md]: Yes, done.

**Comment by scottrigby** [comment]
We have some good user validation of the steps, so let's merge this. It just needs conflicts to be resolved.

**Review by scottrigby — APPROVED** [review]

File: charts/kube-prometheus-stack/README.md
Comment (scottrigby) [reply charts/kube-prometheus-stack/README.md:239]: It sounds like this has been discussed below, but if there's a recommendation for mitigating this further let's update these instructions in a follow-up PR. Right now, there is no good documented migration path so this is an improvement.

**Comment by scottrigby** [comment]
Fixed markdownlint on your branch @fktkrt

**Review by scottrigby — APPROVED** [review]

**Comment by llamahunter** [comment]
The prometheus.prometheusSpec.serviceMonitorSelector changes with the new chart from `release: prometheus-operator` to `release: kube-prometheus-stack`.  If you have a lot of ServiceMonitor objects around, do you then need to go through and manually update all your existing ServiceMonitor objects? Is there any way to make it continue to use the old selector? Looking at the helm chart, it didn't seem possible, as the Release.Name is hard coded into the ServiceMonitor objects created by kube-prometheus-stack.

**Comment by fktkrt** [comment]
I had no issues regarding this with the following setup:

I am using ServiceMonitors and PodMonitors, and labeling those this way:

```
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: foo-podmonitor
spec:
  selector:
    matchLabels:
      app: foo
```

This is my prometheusSpec:

```
prometheus:
  prometheusSpec:
    ruleSelectorNilUsesHelmValues: false
    serviceMonitorSelectorNilUsesHelmValues: false
    podMonitorSelectorNilUsesHelmValues: false
```

additionally, for my custom environments I am overriding these with externalLabels like this:

```
prometheus.prometheusSpec.externalLabels.prometheus=dev-prometheus
```

My Prometheus instance has default labels:

```
Labels:       app=prometheus
              controller-revision-hash=prometheus-kube-prometheus-stack-prometheus-677954c66f
              prometheus=kube-prometheus-stack-prometheus
              statefulset.kubernetes.io/pod-name=prometheus-kube-prometheus-stack-prometheus-0
```

and the new release can continue scraping all my existing Pod/ServiceMonitors created with the legacy chart.

If you have access to a test cluster (or can deploy a playground with k3s or similar), you can easily doublecheck it, just to make sure nothing breaks.

**Comment by fktkrt** [comment]
You can use additional labels to tie apps to specific prometheus instances.
Another idea that just came up is that if you're using additional charts to manage the CRDs, or you can solve this by modifying only your template of your CRD generator chart.

(By the way, you were close with promethei, https://prometheus.io/docs/introduction/faq/#what-is-the-plural-of-prometheus :) )

**Pull request merged**
